File: java/src/main/java/shop/rates/RatesClient.java
package shop.rates;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * The client a stub is put in front of. Its timeout, retry and error
 * mapping all run for real against a programmable server — which is what
 * separates this from mocking the client itself.
 * https://endtoendtester.com/tools/wiremock
 */
public final class RatesClient {
    private static final Pattern GBP = Pattern.compile("\"GBP\"\\s*:\\s*([0-9.]+)");

    private final String baseUrl;
    private final int retries;
    private final HttpClient http;

    public RatesClient(String baseUrl, Duration timeout, int retries) {
        this.baseUrl = baseUrl;
        this.retries = retries;
        this.http = HttpClient.newBuilder().connectTimeout(timeout).build();
    }

    public double gbpRate() {
        RuntimeException last = new RatesUnavailableException("never attempted");

        for (int attempt = 0; attempt <= retries; attempt++) {
            try {
                HttpRequest request = HttpRequest.newBuilder(URI.create(baseUrl + "/v1/rates?base=USD"))
                        .header("Accept", "application/json")
                        .timeout(Duration.ofMillis(500))
                        .build();
                HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 429) {
                    // Not retried: the server said how long to wait.
                    throw new RateLimitedException(
                            Integer.parseInt(response.headers().firstValue("retry-after").orElse("60")));
                }
                if (response.statusCode() != 200) {
                    last = new RatesUnavailableException("upstream returned " + response.statusCode());
                    continue;
                }

                Matcher matcher = GBP.matcher(response.body());
                if (!matcher.find()) {
                    throw new RatesUnavailableException("response did not contain a GBP rate");
                }
                return Double.parseDouble(matcher.group(1));
            } catch (RateLimitedException e) {
                throw e;
            } catch (IOException | InterruptedException e) {
                last = new RatesTimeoutException(e.getMessage() == null ? "no response" : e.getMessage());
            }
        }

        throw last;
    }

    public static class RatesUnavailableException extends RuntimeException {
        public RatesUnavailableException(String message) {
            super(message);
        }
    }

    public static class RatesTimeoutException extends RuntimeException {
        public RatesTimeoutException(String message) {
            super(message);
        }
    }

    public static class RateLimitedException extends RuntimeException {
        private final int retryAfterSeconds;

        public RateLimitedException(int retryAfterSeconds) {
            super("rate limited for " + retryAfterSeconds + "s");
            this.retryAfterSeconds = retryAfterSeconds;
        }

        public int retryAfterSeconds() {
            return retryAfterSeconds;
        }
    }
}

File: typescript/src/testing-levels/integration-testing-with-stubs/rates-client.ts
/* The client under test. A stub replaces the far side of the socket, so
   everything here — serialization, timeouts, retries, error mapping — runs
   for real. Mocking this class instead would skip all of it, which is
   where the bugs are.
   https://endtoendtester.com/testing-levels/integration-testing-with-stubs */

export class RatesUnavailableError extends Error {}
export class RatesTimeoutError extends Error {}
export class RateLimitedError extends Error {
  constructor(readonly retryAfterSeconds: number) {
    super(`rate limited for ${retryAfterSeconds}s`);
  }
}

export interface RatesClientOptions {
  timeoutMs?: number;
  retries?: number;
}

export class RatesClient {
  readonly attempts: number[] = [];

  constructor(
    private readonly baseUrl: string,
    private readonly options: RatesClientOptions = {}
  ) {}

  async gbpRate(): Promise<number> {
    const timeoutMs = this.options.timeoutMs ?? 1_000;
    const retries = this.options.retries ?? 0;
    let lastError: Error = new RatesUnavailableError('never attempted');

    for (let attempt = 0; attempt <= retries; attempt++) {
      this.attempts.push(Date.now());
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeoutMs);

      try {
        const response = await fetch(`${this.baseUrl}/v1/rates?base=USD`, {
          signal: controller.signal,
          headers: { Accept: 'application/json' }
        });

        if (response.status === 429) {
          // Not retried: the server has told us how long to wait, and
          // hammering it is how a rate limit becomes a ban.
          throw new RateLimitedError(Number(response.headers.get('retry-after') ?? 60));
        }
        if (!response.ok) {
          lastError = new RatesUnavailableError(`upstream returned ${response.status}`);
          continue;
        }

        const body = (await response.json()) as { rates?: { GBP?: number } };
        const rate = body.rates?.GBP;
        if (typeof rate !== 'number') {
          throw new RatesUnavailableError('response did not contain a GBP rate');
        }
        return rate;
      } catch (error) {
        if (error instanceof RateLimitedError) throw error;

        /* Detecting a timeout is not as simple as `error.name === 'AbortError'`.
           Node's fetch (undici) wraps the abort: what reaches here is often a
           plain `TypeError: fetch failed` whose `cause` is the AbortError. The
           signal itself is the reliable witness, and this is exactly the kind
           of thing only a test against a real socket finds. */
        if (controller.signal.aborted) {
          lastError = new RatesTimeoutError(`no response within ${timeoutMs}ms`);
          continue;
        }
        lastError = error instanceof Error ? error : new RatesUnavailableError(String(error));
      } finally {
        clearTimeout(timer);
      }
    }

    throw lastError;
  }
}
